plugins/out_clp_s3/flush/flush.go

```diff
@@ -1,88 +1,232 @@
-// Package implements methods to send data to output. All data provided by fluent-bit is encoded
-// with msgpack.
+// Package implements methods to send data to output. All data provided by Fluent Bit is encoded
+// with Msgpack.
 
 package flush
 
 import (
 	"C"
-	"bufio"
+	"encoding/json"
 	"fmt"
+	"io"
+	"log"
 	"os"
 	"path/filepath"
 	"time"
 	"unsafe"
 
 	"github.com/fluent/fluent-bit-go/output"
+	"github.com/klauspost/compress/zstd"
+	"github.com/y-scope/clp-ffi-go/ffi"
+	"github.com/y-scope/clp-ffi-go/ir"
 
-	"github.com/y-scope/fluent-bit-clp/config"
+	"github.com/y-scope/fluent-bit-clp/internal/config"
+	"github.com/y-scope/fluent-bit-clp/internal/decoder"
 )
 
-// Flushes data to file.
+// Flushes data to a file in IR format. Decode of Msgpack based on [Fluent Bit reference].
 //
 // Parameters:
-//   - data: msgpack data
+//   - data: Msgpack data
 //   - length: Byte length
-//   - tag: fluent-bit tag
-//   - S3Config: Configuration based on fluent-bit.conf
+//   - tag: Fluent Bit tag
+//   - S3Config: Plugin configuration
 //
 // Returns:
+//   - code: Fluent Bit success code (OK, RETRY, ERROR)
 //   - err: Error if flush fails
-func File(data unsafe.Pointer, length int, tag string, config *config.S3Config) error {
-	fullFilePath := filepath.Join(config.Path, config.File)
+//
+// [Fluent Bit reference]: https://github.com/fluent/fluent-bit-go/blob/a7a013e2473cdf62d7320822658d5816b3063758/examples/out_multiinstance/out.go#L41
+func ToFile(data unsafe.Pointer, length int, tag string, config *config.S3Config) (int, error) {
+	// Buffer to store events from Fluent Bit chunk.
+	var logEvents []ffi.LogEvent
 
-	// If the file doesn't exist, create it, or append to the file. Will still cause error if there
-	// is no directory
-	f, err := os.OpenFile(fullFilePath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0o644)
-	if err != nil {
-		return err
+	dec := decoder.New(data, length)
+
+	// Loop through all records in Fluent Bit chunk.
+	for {
+		ts, record, err := decoder.GetRecord(dec)
+		if err != nil {
+			if err == io.EOF {
+				// Chunk decoding finished. Break out of loop and send log events to output.
+				break
+			} else {
+				err = fmt.Errorf("error decoding data from stream: %w", err)
+				return output.FLB_ERROR, err
+			}
+		}
```

Comment:
Lets simplify this a little.
```suggestion
		if err == io.EOF {
			// Chunk decoding finished. Break out of loop and send log events to output.
			break
		} else if err != nil {
			err = fmt.Errorf("error decoding data from stream: %w", err)
			return output.FLB_ERROR, err
		}
```